Elly Vending step8

dely2p opened the pull request:
Vending step8 입니다.

dely2p committed:
add PieGraphView class that inherits the UIView #56

UIView를 상속받는 PieGraphView 클래스를 추가함

dely2p committed:
add PieDrawable protocol to get list of beverage purchases #56

음료 구매 목록을 가져오기 위해서 PieDrawable 프로토콜 추가

dely2p committed:
fix PieGraphView to draw pie graph by stroke #57

stroke로 pie graph 그리기를 구현함

dely2p committed:
fix code from stroke to fill #57

선으로 그렸던 것을 색 채우기로 수정했다.

dely2p committed:
fix bug in pie graph shape#57

파이그래프 모양이 잘못나오는 버그를 고침

dely2p committed:
add text on pie graph #57

파이그래프 위에 각 음료의 이름이 나타나도록 구현함

godrm commented on VendingMachineApp/VendingMachineApp/Model/PieGraphView.swift:
```diff
@@ -0,0 +1,93 @@
+    }
+    var pieDrawable: PieDrawable? {
+        didSet {
+            guard let data = pieDrawable?.sendData() else {
```
sendData()는 어디서 어디로 send를 하는걸까요? 
ImagePickerControllerDelegate를 떠올려보세요...

godrm commented on VendingMachineApp/VendingMachineApp/Model/PieGraphView.swift:
```diff
@@ -0,0 +1,93 @@
+
+            textToRender.draw(in: renderRect, withAttributes: textAttributes)
+            
+            startAngle = endAngle
```
위에서는 함수로 작게 쪼개놨는데, 실제로 그리는 함수는 조금 기네요. 
역할 단위로 메소드를 분리해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/PieGraphView.swift:
```diff
@@ -0,0 +1,93 @@
+    override func draw(_ rect: CGRect) {
+        // 중심점과 반지름 계산
+        let center = CGPoint(x: bounds.width / 2, y: bounds.height / 2)
+        let radius: CGFloat = max(bounds.width, bounds.height)/2 - Constants.arcWidth/2
```
draw() 함수는 드로잉 엔진에 따라서 엄청 자주 불립니다. 
let 상수더라도 항상 같은 값이면 인스턴스 상수로 표현하는 게 좋습니다.

godrm reviewed (commented):
피드백 확인해보세요.

dely2p committed:
fix constant position from draw() method to constant instance #61

draw 내에 선언한 상수를 인스턴스 상수로 수정함

dely2p committed:
fix draw method to separate functions #60

draw() method 내의 여러기능들을 각 메소드로 분리함

dely2p committed:
fix sendData method name to receiveData method #59

메소드명의 모호성으로 PieViewGraph에서 ManagerViewController로부터 받아오는 데이터를 의미하는 receiveData()로 메소드명 개선함

dely2p committed:
remove print code to test #59

동적으로 값이들어가는 프로퍼티 테스트용 print() 제거

dely2p commented:
피드백 주신 부분 수정했습니다:)

- ManagerViewController에서 PieGraphView로 데이터를 넘겨주기 때문에 receiveData()로 메소드명을 변경했습니다.
- draw() 내의 코드들을 setPieGraphdrawing(), drawPieGraph(), drawTextOnPieGraph()로 메소드 분리를 했습니다.
- draw() 내의 let 상수들을 인스턴스 상수로 수정했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/PieGraphView.swift:
```diff
@@ -26,13 +26,20 @@ class PieGraphView: UIView {
+        return CGPoint(x: bounds.width / 2, y: self.bounds.height / 2)
+    }
+    private var radiusOfGraph: CGFloat {
+        return max(bounds.width, bounds.height)/2 - Constants.arcWidth/2
```
computed property로 만드는 건 잘했어요
그런데 이런 값들은 View가 만들어지고 나면 바뀌지 않는 상수에 가깝습니다.
계속 계산하는 대신 한 번만 계산하고 같은 값을 계속 사용해도 된다는거죠.
참고하세요.

godrm commented:
마지막 마무리하고 다음 미션을 진행합시다.
